rocAL/source/readers/image/caffe2_lmdb_record_reader.cpp

```diff
@@ -35,58 +36,88 @@ Caffe2LMDBRecordReader::Caffe2LMDBRecordReader() {
+        // which equals to the shard size with padding
+        if (_last_batch_info.last_batch_policy == RocalBatchPolicy::PARTIAL || _last_batch_info.last_batch_policy == RocalBatchPolicy::FILL) {
+            ret += _last_batch_padded_size;
+        } else if (_last_batch_info.last_batch_policy == RocalBatchPolicy::DROP &&
+                   _last_batch_info.pad_last_batch_repeated == true) {  // When pad_last_batch_repeated is False - Enough
+                                                       // number of samples would not be present in the last batch - hence
+                                                       // dropped by condition handled in the loader
+            ret -= _batch_size;
+        }
+    } else if (_shard_size > 0) {
+        auto largest_shard_size_with_padding =
+            _shard_size + (_batch_size - (_shard_size % _batch_size));  // The shard size used here is padded
+        if (_loop)
+            return largest_shard_size_with_padding;
+        int size = std::max(largest_shard_size_with_padding, _batch_size);
+        ret = (size - _read_counter);
+        if (_last_batch_info.last_batch_policy == RocalBatchPolicy::DROP)  // The shard size is padded at the beginning of the condition, hence dropping the last batch
+            ret -= _batch_size;
+    }
     return ((ret < 0) ? 0 : ret);
 }
 
 Reader::Status Caffe2LMDBRecordReader::initialize(ReaderConfig desc) {
     auto ret = Reader::Status::OK;
-    _file_id = 0;
     _folder_path = desc.path();
     _path = desc.path();
     _shard_id = desc.get_shard_id();
     _shard_count = desc.get_shard_count();
-    _batch_count = desc.get_batch_size();
     _loop = desc.loop();
     _shuffle = desc.shuffle();
+    _batch_size = desc.get_batch_size();
+    _last_batch_info = desc.get_sharding_info();
+    _pad_last_batch_repeated = _last_batch_info.pad_last_batch_repeated;
+    _stick_to_shard = _last_batch_info.stick_to_shard;
+    _shard_size = _last_batch_info.shard_size;
     ret = folder_reading();
-    // the following code is required to make every shard the same size:: required for multi-gpu training
-    if (_shard_count > 1 && _batch_count > 1) {
-        int _num_batches = _file_names.size() / _batch_count;
-        int max_batches_per_shard = (_file_count_all_shards + _shard_count - 1) / _shard_count;
-        max_batches_per_shard = (max_batches_per_shard + _batch_count - 1) / _batch_count;
-        if (_num_batches < max_batches_per_shard) {
-            replicate_last_batch_to_pad_partial_shard();
-        }
-    }
+    _curr_file_idx = _shard_start_idx_vector[_shard_id]; // shard's start_idx would vary for every shard in the vector
     // shuffle dataset if set
     if (ret == Reader::Status::OK && _shuffle)
-        std::random_shuffle(_file_names.begin(), _file_names.end());
+        std::random_shuffle(_all_shard_file_names_padded.begin() + _shard_start_idx_vector[_shard_id],
```

Comment:
File source reader does not have this _all_shard_file_names_padded and is replaced with _file_names

Please make all changes similar to file source reader for all the readers